src/Interpreters/InterpreterInsertQuery.cpp

```diff
@@ -758,6 +761,12 @@ InterpreterInsertQuery::distributedWriteIntoReplicatedMergeTreeFromClusterStorag
     {
         if (auto * select_query = select.list_of_selects->children.at(0)->as<ASTSelectQuery>())
         {
+            if (local_context->getSettingsRef()[Setting::enable_global_with_statement])
+            {
+                ApplyWithAliasVisitor::visit(select.list_of_selects->children.at(0));
+                ApplyWithSubqueryVisitor(local_context).visit(select.list_of_selects->children.at(0));
```

Comment:
In the [InterpreterSelectQuery](https://github.com/ClickHouse/ClickHouse/blob/48e027c8ba999ea866cd507d1018bfc4c4ab9d73/src/Interpreters/InterpreterSelectQuery.cpp#L547), we call this visitor outside the `if`